proxy/http/remap/NextHopConfig.cc

```diff
@@ -0,0 +1,387 @@
+/** @file
+
+  A brief file description
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include "NextHopConfig.h"
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <unordered_set>
+#include <string.h>
+
+#include "ts/Diags.h"
+
+#define BUFSIZE 8192
+
+ts::Errata
+NextHopConfig::loadConfig(const char *fileName)
+{
+  std::stringstream doc;
+  std::unordered_set<std::string> include_once;
+
+  try {
+    loadFile(fileName, doc, include_once);
+    YAML::Node node = YAML::Load(doc);
+    config          = node.as<NextHopStrategyConfig>();
+
+  } catch (std::exception &ex) {
+    errata.push(ts::Errata::Message(1, 1, ex.what()));
+    return errata;
+  }
+  return errata;
+}
+
+void
+NextHopConfig::loadFile(const std::string fileName, std::stringstream &doc, std::unordered_set<std::string> &include_once)
+{
+  const char *sep = " \t";
+  char *tok, *last;
+  char line[BUFSIZE];
+
+  std::ifstream file(fileName);
+  if (file.is_open()) {
+    while (file.getline(line, BUFSIZE - 1)) {
+      if (line[0] == '#') {
+        tok = strtok_r(line, sep, &last);
+        if (tok != nullptr && strcmp(tok, "#include") == 0) {
+          std::string f = strtok_r(nullptr, sep, &last);
+          if (include_once.find(f) == include_once.end()) {
+            include_once.insert(f);
+            try {
+              loadFile(f, doc, include_once);
+            } catch (std::exception &ex) {
```

Comment:
What is the purpose of this?

Reply:
if there is a '#include hosts.yaml' then that file is read into the doc at that point, using recursive call to 'loadFile()'. include_once is used to make sure the file is only included once into a yaml document.

So, we may have a single hosts.yaml with all the host definitions but multiple strategies.  Since yaml does not support anchors accross documents or support importing other yaml files, i implemented an 'include'.

Reply:
No, I meant why `catch` and rethrow the exact same exception?